frontend/src/uxbox/main/data/workspace.cljs

```diff
@@ -598,6 +603,30 @@
       (let [pid (:current-page-id state)]
         (update-in state [:workspace-data pid :objects id] merge attrs)))))
 
+(defn update-shape-recursive
+  [id attrs]
+  (us/verify ::us/uuid id)
+  (us/verify ::shape-attrs attrs)
+  (ptk/reify ::update-shape
+    dwc/IBatchedChange
+    dwc/IUpdateGroup
+    (get-ids [_] [id])
+
+    ptk/UpdateEvent
+    (update [_ state]
+      (let [page-id (:current-page-id state)
+            grouped #{:frame :group}]
+        (update-in state [:workspace-data page-id :objects]
+                   (fn [objects]
+                     (let [childs (d/concat [id] (helpers/get-children id objects))]
```

Comment:
`helpers/get-children` is already returning all recursive children so you don't really need to go recursive, just iterate through children.